lsat/interceptor.go

```diff
@@ -0,0 +1,184 @@
+	req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker,
+	opts ...grpc.CallOption) error {
+
+	addLsatCredentials := func(token *Token) error {
+		macaroon, err := token.PaidMacaroon()
+		if err != nil {
+			return err
+		}
+		opts = append(opts, grpc.PerRPCCredentials(
+			macaroons.NewMacaroonCredential(macaroon),
+		))
+		return nil
+	}
+
+	// If we already have a token, let's append it.
+	if i.store.HasToken() {
+		lsat, err := i.store.Token()
+		if err != nil {
+			return err
+		}
+		if err = addLsatCredentials(lsat); err != nil {
+			return err
+		}
+	}
+
+	// We need a way to extract the response headers sent by the
+	// server. This can only be done through the experimental
+	// grpc.Trailer call option.
+	// We execute the request and inspect the error. If it's the
+	// LSAT specific payment required error, we might execute the
+	// same method again later with the paid LSAT token.
+	trailerMetadata := &metadata.MD{}
+	opts = append(opts, grpc.Trailer(trailerMetadata))
+	err := invoker(ctx, method, req, reply, cc, opts...)
+
+	// Only handle the LSAT error message that comes in the form of
+	// a gRPC status error.
+	if isPaymentRequired(err) {
+		lsat, err := i.payLsatToken(trailerMetadata)
+		if err != nil {
+			return err
+		}
+		if err = addLsatCredentials(lsat); err != nil {
+			return err
+		}
+
+		// Execute the same request again, now with the LSAT
+		// token added as an RPC credential.
+		return invoker(ctx, method, req, reply, cc, opts...)
+	}
+	return err
+}
+
+// payLsatToken reads the payment challenge from the response metadata and tries
+// to pay the invoice encoded in them, returning a paid LSAT token if
+// successful.
+func (i *Interceptor) payLsatToken(md *metadata.MD) (*Token, error) {
+	// First parse the authentication header that was stored in the
+	// metadata.
+	authHeader := md.Get(AuthHeader)
+	if len(authHeader) == 0 {
+		return nil, fmt.Errorf("auth header not found in response")
+	}
+	matches := authHeaderRegex.FindStringSubmatch(authHeader[0])
+	if len(matches) != 3 {
+		return nil, fmt.Errorf("invalid auth header "+
+			"format: %s", authHeader[0])
+	}
+
+	// Decode the base64 macaroon and the invoice so we can store the
+	// information in our store later.
+	macBase64, invoiceStr := matches[1], matches[2]
+	macBytes, err := base64.StdEncoding.DecodeString(macBase64)
+	if err != nil {
+		return nil, fmt.Errorf("base64 decode of macaroon failed: "+
+			"%v", err)
+	}
+	invoice, err := zpay32.Decode(invoiceStr, i.lnd.ChainParams)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode invoice: %v", err)
+	}
+	token, err := NewToken(macBytes, invoice.PaymentHash)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create token: %v", err)
+	}
+
+	// Pay invoice now and wait for the result to arrive.
+	respChan := i.lnd.Client.PayInvoice(
+		context.Background(), invoiceStr, MaxRoutingFee, nil,
+	)
+	result := <-respChan
```

Comment:
Should pass in the context and also select on its `Done()` channel so this can be properly cancelled.